### PR TITLE
Remove strftime (deprecated in php 8.1)

### PR DESCRIPTION
**Pull request by sveneld**
Add possibility to configure timeFormatter

### CODE REVIEW

File: Log.php [hunk 1]
```diff
@@ -869,4 +869,64 @@ function getIdent()
...
+     * @param string $timeFormat
+     * @param int|null $time
+     * @param callable|null $timeFormatter
+     * @return string
+     */
+    public function timeFormat($timeFormat = 'Y-m-d H:i:s', $time = null, callable $timeFormatter = null)
+    {
+        if (is_null($time)) {
+            $time = time();
+        }
```
Comment (jparise) [inline Log.php:883]: I think the caller should always pass a `$time` value (i.e. it shouldn't be nullable). In practice, defaulting it to `null` is just moving all of the `time()` calls inside of this formatting function, and I think it would be better if that value was explicitly provided.

File: Log.php [hunk 2]
```diff
@@ -869,4 +869,64 @@ function getIdent()
     {
         return $this->_ident;
     }
+
+    /**
```
Comment (jparise) [inline Log.php:873]: Could you add a little bit more of a documentation comment here?

File: Log.php [hunk 3]
```diff
@@ -869,4 +869,64 @@ function getIdent()
...
+    {
+        if (is_null($time)) {
+            $time = time();
+        }
+        if (!is_null($timeFormatter) && is_callable($timeFormatter)) {
+            return call_user_func($timeFormatter, $timeFormat, $time);
+        }
+
+        if (strstr($timeFormat, '%') !== false) {
+            trigger_error('Using timeFormat string for strftime is deprecated', E_USER_WARNING);
```
Comment (jparise) [inline Log.php:889]: Maybe rephrase this to say something like "strftime-style formatting"? The end user isn't necessarily aware of the internal use of the `strftime()` function.

File: Log.php [hunk 4]
```diff
@@ -869,4 +869,64 @@ function getIdent()
...
+            $timeFormat = $this->convertStrftimeFormatConverter($timeFormat);
+        }
+
+        return date($timeFormat, $time);
+    }
+
+    /**
```
Comment (jparise) [inline Log.php:896]: This could also use a descriptive documentation comment.

File: Log.php [hunk 5]
```diff
@@ -869,4 +869,64 @@ function getIdent()
...
+    public function timeFormat($timeFormat = 'Y-m-d H:i:s', $time = null, callable $timeFormatter = null)
+    {
+        if (is_null($time)) {
+            $time = time();
+        }
+        if (!is_null($timeFormatter) && is_callable($timeFormatter)) {
+            return call_user_func($timeFormatter, $timeFormat, $time);
+        }
+
+        if (strstr($timeFormat, '%') !== false) {
```
Comment (jparise) [inline Log.php:888]: I think [`strpos()`](https://www.php.net/manual/en/function.strpos.php) might be a little more efficient for a "does this string contain a `%`" test.

**Review by jparise — COMMENTED** [review]
This looks good. Thanks!

In addition to the comments I left, the `docs/guide.txt` documentation will also need a few updates.

File: Log.php (hunk 5 above)
Comment (sveneld) [reply Log.php:888]: Agree :)

File: Log.php (hunk 4 above)
Comment (sveneld) [reply Log.php:896]: ++

File: Log.php (hunk 3 above)
Comment (sveneld) [reply Log.php:889]: ++

File: Log.php (hunk 2 above)
Comment (sveneld) [reply Log.php:873]: ++

File: Log.php (hunk 1 above)
Comment (sveneld) [reply Log.php:883]: ++

**Comment by sveneld** [comment]
modified pull request by your suggestion

File: Log.php [hunk 6]
```diff
@@ -869,4 +871,69 @@ function getIdent()
...
     }
+
+    /**
+     * Function to format time in specified format, which will be used in log record
+     * By default will be used format self::TIME_FORMAT
+     * timeFormatter function will be used if it is set
+     *
+     * @param string $timeFormat
+     * @param int $time
+     * @param callable|null $timeFormatter
```
Comment (jparise) [inline Log.php:882]: I think some of these phrases would be better as `@param` documentation comments. Move them down there?

(Minor feedback, but this is a good time to apply it.)

File: Log.php [hunk 7]
```diff
@@ -869,4 +871,69 @@ function getIdent()
...
+     * @return string
+     */
+    public function timeFormat($timeFormat, $time, callable $timeFormatter = null)
```
Comment (jparise) [inline Log.php:885]: I think I have a small preference for `$time` now being the first argument.

And them maybe rename this function to `formatTime`?

File: Log.php [hunk 8]
```diff
@@ -37,6 +37,8 @@
  */
 class Log
 {
+    const TIME_FORMAT = 'M d H:i:s';
```
Comment (jparise) [inline Log.php:40]: Do you think this should be `DEFAULT_TIME_FORMAT`, given how it's used below?

File: Log.php [hunk 9]
```diff
@@ -869,4 +871,69 @@ function getIdent()
...
+    public function timeFormat($timeFormat, $time, callable $timeFormatter = null)
+    {
+        $timeFormat = empty($timeFormat) ? self::TIME_FORMAT : $timeFormat;
+
+        if (!is_null($timeFormatter) && is_callable($timeFormatter)) {
+            return call_user_func($timeFormatter, $timeFormat, $time);
+        }
+
+        if (strpos($timeFormat, '%') !== false) {
+            trigger_error('Using strftime style formatting is deprecated', E_USER_WARNING);
```
Comment (jparise) [inline Log.php:894]:
```suggestion
            trigger_error('Using strftime-style formatting is deprecated', E_USER_WARNING);
```

File: docs/guide.txt [hunk 10]
```diff
@@ -322,11 +322,14 @@ Configuration
 |                   |           | [%3$s] %4$s`` | specification.            |
 +-------------------+-----------+---------------+---------------------------+
 | ``timeFormat``    | String    | ``%b %d       | Time stamp format         |
-|                   |           | %H:%M:%S``    | (for strftime_).          |
+|                   |           | %H:%M:%S``    | (for date_).              |
```
Comment (jparise) [inline docs/guide.txt:325]: The `Default` column value (here and below) should also be updated with the new `data()`-style format value.

**Review by jparise — COMMENTED** [review]
It would also nice to have a unit test for the new formatting behavior (primarily the format converter).

File: docs/guide.txt (hunk 10 above)
Comment (sveneld) [reply docs/guide.txt:325]: Sorry, forget to change it.

File: Log.php (hunk 6 above)
Comment (sveneld) [reply Log.php:882]: ++

File: Log.php (hunk 7 above)
Comment (sveneld) [reply Log.php:885]: ++

File: Log.php (hunk 8 above)
Comment (sveneld) [reply Log.php:40]: ++

File: Log.php (hunk 9 above)
Comment (sveneld) [reply Log.php:894]: ++

**Comment by sveneld** [comment]
Add some tests

**Review by jparise — APPROVED** [review]
Nice work! Will merge after the conflicts from #24 are resolved.

**Comment by sveneld** [comment]
Done
